From ca8cc41ad36d18f869454dbfdabddd07d32cb398 Mon Sep 17 00:00:00 2001
From: Jean Christophe Beyler <jean.christophe.beyler@intel.com>
Date: Fri, 8 Feb 2013 07:11:18 -0800
Subject: Dalvik: Growable List is not correct when initial size is 0

BZ: 86041

Fixes the expansion to generically handle the case where the initial size was 0

Utility.cpp:
    - Instead of doing oldSize * 2, we do oldSize * 2 + 1 to ensure expansion
      for the 0 case

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I6f97b593e1c30d8d0a15edc4bca25b599a87e6d5
Orig-MCG-Change-Id: Ief67a0ce6cd0942b95acdc1aa3078b63569aa4ed
Signed-off-by: Jean Christophe Beyler <jean.christophe.beyler@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/Utility.cpp |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/vm/compiler/Utility.cpp b/vm/compiler/Utility.cpp
index ecfb85d..85018dc 100644
--- a/vm/compiler/Utility.cpp
+++ b/vm/compiler/Utility.cpp
@@ -229,6 +229,8 @@ static void expandGrowableList(GrowableList *gList)
     int newLength = gList->numAllocated;
     if (newLength < 128) {
         newLength <<= 1;
+        //If ever newLength was set to 0, we add 1 to at least expand it
+        newLength += 1;
     } else {
         newLength += 128;
     }
-- 
1.7.4.1

